### PR TITLE
validator txn api fix

### PR DESCRIPTION
**Pull request by zjma**
### TLDR

-  For internal txn type `ValidatorTransaction::JWKUpdate`.
  - New JSON property `validator_transaction_type`, value being `"observed_jwk_update"`.
  - New JSON property  `quorum_certified_update`: details of the JWK updates.
- For internal txn type `ValidatorTransaction::DKGResult`.
  - New JSON property `validator_transaction_type`, value being `"dkg_result"`.
  - New JSON property `dkg_transcript`: details of the DKG transcript.
- The following events are moved from `BlockMetadataTransaction` to `ValidatorTransaction::DKGResult` when randomness is enabled.
  - `0x1::stake::DistributeRewardsEvent`
  - `0x1::stake::DistributeRewards`
  - `0x1::reconfiguration::NewEpoch`
  - `0x1::reconfiguration::NewEpochEvent`

### Example JSONs

example  `ValidatorTransaction::JWKUpdate` json:
```json
{
    "version": "2",
    "hash": "0xc52861c077546788b0cdd571935ef70de331b78411313e273f1a4c6110797342",
    "state_change_hash": "0xc5ad04ecdae5603876c077a635650195d2a166309894094b68e9630847a6da31",
    "event_root_hash": "0xd256b1260508902acf6ea7592fe8da54a76d9759a9f34569d77273302637329d",
    "state_checkpoint_hash": null,
    "gas_used": "0",
    "success": true,
    "vm_status": "Executed successfully",
    "accumulator_root_hash": "0xc630139580da3da1b896e690b9d4b558d77ac4950eb0a1e0236e1b9cc2a1abd6",
    "changes": [], //omitted
    "events": [], //omitted
    "timestamp": "1718424440685360",
    "validator_transaction_type": "observed_jwk_update", // new property!
    "type": "validator_transaction"
}
```

example `ValidatorTransaction::DKGResult` json:
```json
{
    "version": "9",
    "hash": "0x0bff715ffcf24ec4c0fb504da45059104774d52a1bcdc5eef1e574fcc6dd6840",
    "state_change_hash": "0xc74bb84ae5f647c60022d48de9e6bed5bb2e8dac6c8cd9b6ca5e26b3edf1cfbd",
    "event_root_hash": "0xe4e0955b86dad038a8dd96fe1ed67b1c893500d2e825e7fb41de4c0fa2325ae5",
    "state_checkpoint_hash": "0xef943d37d839f33b0416366b453ec682ef258ce1a08c5195e711277b7a32bbcd",
    "gas_used": "0",
    "success": true,
    "vm_status": "Executed successfully",
    "accumulator_root_hash": "0x4e579a1e2434bde316d597e8f2dd621371a95524891b0b7b44f69671d19bfef8",
    "changes": [], //omitted
    "events": [ // these are in BlockMetadataTransaction when randomness is disabled!
        {
            "guid": {
                "creation_number": "0",
                "account_address": "0x0"
            },
            "sequence_number": "0",
            "type": "0x1::stake::DistributeRewards",
            "data": {
                "pool_address": "0xdf8211644b8da986605fd92f43d3fc823f49fff21891e78871ccfdb0d1608f47",
                "rewards_amount": "0"
            }
        },
        {
            "guid": {
                "creation_number": "12",
                "account_address": "0xdf8211644b8da986605fd92f43d3fc823f49fff21891e78871ccfdb0d1608f47"
            },
            "sequence_number": "0",
            "type": "0x1::stake::DistributeRewardsEvent",
            "data": {
                "pool_address": "0xdf8211644b8da986605fd92f43d3fc823f49fff21891e78871ccfdb0d1608f47",
                "rewards_amount": "0"
            }
        },
        {
            "guid": {
                "creation_number": "0",
                "account_address": "0x0"
            },
            "sequence_number": "0",
            "type": "0x1::reconfiguration::NewEpoch",
            "data": {
                "epoch": "2"
            }
        },
        {
            "guid": {
                "creation_number": "2",
                "account_address": "0x1"
            },
            "sequence_number": "1",
            "type": "0x1::reconfiguration::NewEpochEvent",
            "data": {
                "epoch": "2"
            }
        }
    ],
    "timestamp": "1718424443766159",
    "validator_transaction_type": "dkg_result", // new property!
    "type": "validator_transaction"
}
```

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs [hunk 1]
```diff
@@ -760,6 +760,9 @@ pub fn convert_transaction(
         Transaction::ValidatorTransaction(_) => {
             transaction::transaction::TransactionType::Validator
         },
+        Transaction::BlockMetadataExtTransaction(_) => {
+            todo!()
```
Comment (gregnazario) [inline ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs:764]: These will be needed for indexing I think?

File: api/types/src/transaction.rs [hunk 2]
```diff
@@ -586,13 +570,278 @@ pub struct BlockMetadataTransaction {
...
+
+impl ExportedQuorumCertifiedUpdate {
+    pub fn from_internal_repr(internal: QuorumCertifiedUpdate) -> Self {
+        let QuorumCertifiedUpdate { update, multi_sig } = internal;
+        Self {
+            update: ExportedProviderJWKs::from_internal_repr(update),
+            multi_sig: ExportedAggregateSignature::from_internal_repr(multi_sig),
+        }
+    }
+}
```
Comment (gregnazario) [inline api/types/src/transaction.rs:778]: This is usually best represented in Rust as a straight implements From

File: api/types/src/transaction.rs [hunk 3]
```diff
@@ -586,13 +570,278 @@ pub struct BlockMetadataTransaction {
...
+    pub round: U64,
+    /// The events emitted at the block creation
     pub events: Vec<Event>,
+    /// Previous block votes
+    pub previous_block_votes_bitvec: Vec<u8>,
+    pub proposer: Address,
+    /// The indices of the proposers who failed to propose
+    pub failed_proposer_indices: Vec<u32>,
     pub timestamp: U64,
+    pub randomness: Option<Vec<u8>>,
```
Comment (gregnazario) [inline api/types/src/transaction.rs:613]: Similar with skip serialization

File: api/types/src/convert.rs [hunk 4]
```diff
@@ -228,7 +237,11 @@ impl<'a, S: StateView> MoveConverter<'a, S> {
                     },
                 })
             },
-            ValidatorTransaction(_txn) => (info, events, timestamp).into(),
+            aptos_types::transaction::Transaction::ValidatorTransaction(txn) => {
+                Transaction::ValidatorTransaction(ValidatorTransaction::from_internal_repr(
```
Comment (gregnazario) [inline api/types/src/convert.rs:241]: Generally I would have taken the approach of special casing the single one rather than changing the rest of the match statement

File: ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs (hunk 1 above)
Comment (zjma) [reply ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs:764]: yes, @bowenyang007 @larry-aptos i assume you guys will have a sub-PR for here?

File: api/types/src/transaction.rs (hunk 2 above)
Comment (zjma) [reply api/types/src/transaction.rs:778]: i found them hard to locate, and the only benefit is `.into()`?

File: api/types/src/convert.rs (hunk 4 above)
Comment (zjma) [reply api/types/src/convert.rs:243]: fixed

File: types/src/block_metadata_ext.rs [hunk 5]
```diff
@@ -104,6 +104,13 @@ impl BlockMetadataExt {
             BlockMetadataExt::V1(obj) => obj.round,
         }
     }
+
+    pub fn type_name(&self) -> &'static str {
+        match self {
+            BlockMetadataExt::V0(_) => "block_metadata_ext_transaction__v0",
+            BlockMetadataExt::V1(_) => "block_metadata_ext_transaction__v1",
```
[the remark this answers is not in the document]
Comment (zjma) [reply types/src/block_metadata_ext.rs:111]: The original is `BlockMetadataExt::V0`, not `BlockMetadataExtV0`, hopefully hinted by double underscoring.

File: api/types/src/transaction.rs [hunk 6]
```diff
@@ -586,13 +570,278 @@ pub struct BlockMetadataTransaction {
...
+            multi_sig: ExportedAggregateSignature::from_internal_repr(multi_sig),
+        }
+    }
+}
+
+/// A more API-friendly representation of the on-chain `aptos_types::aggregate_signature::AggregateSignature`.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Object)]
+pub struct ExportedAggregateSignature {
+    signer_indices: Vec<usize>,
+    sig: Option<Vec<u8>>,
```
[the remark this answers is not in the document]
Comment (zjma) [reply api/types/src/transaction.rs:794]: i saw other `Options` serialized to nulls, e.g. `state_checkpoint_hash: null`. I thought they are more explicit and are good?

File: api/types/src/transaction.rs [hunk 7]
```diff
@@ -354,15 +338,15 @@ impl From<(&SignedTransaction, TransactionPayload)> for UserTransactionRequest {
...
-}
+// impl From<(TransactionInfo, Vec<Event>, u64)> for Transaction {
+//     fn from((info, events, timestamp): (TransactionInfo, Vec<Event>, u64)) -> Self {
+//         Transaction::ValidatorTransaction(ValidatorTransaction {
+//             info,
+//             events,
+//             timestamp: timestamp.into(),
+//         })
+//     }
+// }
```
[the remark this answers is not in the document]
Comment (zjma) [reply api/types/src/transaction.rs:347]: fixed

File: ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs [hunk 8]
```diff
@@ -829,6 +832,9 @@ pub fn convert_transaction(
         Transaction::ValidatorTransaction(_) => {
             transaction::transaction::TxnData::Validator(transaction::ValidatorTransaction {})
         },
+        Transaction::BlockMetadataExtTransaction(_) => {
+            todo!()
```
Comment (zjma) [inline ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs:836]: indexer todo?
@bowenyang007 @ladiJ2N

File: crates/indexer/src/models/transactions.rs [hunk 9]
```diff
@@ -250,6 +250,9 @@ impl Transaction {
                 vec![],
                 vec![],
             ),
+            APITransaction::BlockMetadataExtTransaction(_block_metadata_ext_txn) => {
+                todo!()
```
Comment (zjma) [inline crates/indexer/src/models/transactions.rs:254]: indexer todo?
@bowenyang007 @larry-aptos

File: ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs (hunk 8 above)
Comment (zjma) [reply ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs:836]: @bowenyang007 @larry-aptos

File: api/types/src/transaction.rs (hunk 2 above)
Comment (gregnazario) [reply api/types/src/transaction.rs:778]: It's more appropriately sharable across.

So, you can do both From and Into at any place.  Helps with other things automatically as well.

File: api/types/src/transaction.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (gregnazario) [reply api/types/src/transaction.rs:784]: They were not implemented correctly, and they cause downstream things to have to handle the nulls

[the remark this answers is not in the document]
Comment (zjma) [reply api/types/src/transaction.rs:706]: fixed

File: api/types/src/transaction.rs (hunk 3 above)
Comment (zjma) [reply api/types/src/transaction.rs]: fixed

File: crates/indexer/src/models/transactions.rs (hunk 9 above)
Comment (zjma) [reply crates/indexer/src/models/transactions.rs]: n/a

File: ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs (hunk 8 above)
Comment (zjma) [reply ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs]: n/a

File: ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs (hunk 1 above)
Comment (zjma) [reply ecosystem/indexer-grpc/indexer-grpc-fullnode/src/convert.rs]: n/a

File: types/src/block_metadata_ext.rs (hunk 5 above)
Comment (gregnazario) [inline types/src/block_metadata_ext.rs:111]: remove

File: api/types/src/transaction.rs [hunk 10]
```diff
@@ -586,13 +579,197 @@ pub struct BlockMetadataTransaction {
...
+            epoch: internal.epoch().into(),
+            round: internal.round().into(),
+            events,
+            previous_block_votes_bitvec: internal.previous_block_votes_bitvec().clone(),
+            proposer: internal.proposer().into(),
+            failed_proposer_indices: internal.failed_proposer_indices().clone(),
+            timestamp: internal.timestamp_usecs().into(),
+        }
+    }
+}
```
Comment (gregnazario) [inline api/types/src/transaction.rs:600]: remove

File: api/types/src/transaction.rs [hunk 11]
```diff
@@ -586,13 +579,197 @@ pub struct BlockMetadataTransaction {
...
+            ValidatorTransaction::DkgResult(t) => t.timestamp,
+        }
+    }
+
+    pub fn from_internal_repr(
+        internal: aptos_types::validator_txn::ValidatorTransaction,
+        info: TransactionInfo,
+        events: Vec<Event>,
+        timestamp: u64,
+    ) -> Self {
```
Comment (gregnazario) [inline api/types/src/transaction.rs:650]: It would still be preferred to be using the native From in the language

File: api/types/src/transaction.rs [hunk 12]
```diff
@@ -586,13 +579,197 @@ pub struct BlockMetadataTransaction {
...
+        }
+    }
+}
+
+/// A more API-friendly representation of the on-chain `aptos_types::aggregate_signature::AggregateSignature`.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Object)]
+pub struct ExportedAggregateSignature {
+    signer_indices: Vec<usize>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    sig: Option<Vec<u8>>,
```
Comment (gregnazario) [inline api/types/src/transaction.rs:706]: Also, I all Vec<U8> I think should be using SerdeBytes?  Unsure how this comes out

File: api/types/src/transaction.rs [hunk 13]
```diff
@@ -586,13 +579,197 @@ pub struct BlockMetadataTransaction {
...
     pub events: Vec<Event>,
     pub timestamp: U64,
+    pub dkg_transcript: ExportedDKGTranscript,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Object)]
+pub struct ExportedDKGTranscript {
+    epoch: U64,
+    author: Address,
+    payload: Vec<u8>,
```
Comment (gregnazario) [inline api/types/src/transaction.rs:757]: Similar here

File: api/types/src/transaction.rs (hunk 12 above)
Comment (zjma) [reply api/types/src/transaction.rs:706]: you mean `#[serde(with = "serde_bytes")]` right? it doesn't seem to make a difference. With/without it, the serialized value is like `[0, 1, 2, ...]`

File: api/types/src/transaction.rs (hunk 13 above)
Comment (zjma) [reply api/types/src/transaction.rs:757]: you mean `#[serde(with = "serde_bytes")]` right? it doesn't seem to make a difference. With/without it, the serialized value is like `[0, 1, 2, ...]`

File: api/types/src/transaction.rs (hunk 12 above)
Comment (zjma) [reply api/types/src/transaction.rs]: i think the existing `HexEncodedBytes` helps. updated to that.

File: api/types/src/transaction.rs (hunk 13 above)
Comment (zjma) [reply api/types/src/transaction.rs]: i think the existing `HexEncodedBytes` helps. updated to that.

File: api/types/src/transaction.rs (hunk 11 above)
Comment (zjma) [reply api/types/src/transaction.rs]: fixed

File: api/types/src/transaction.rs (hunk 10 above)
Comment (zjma) [reply api/types/src/transaction.rs]: fixed

File: types/src/block_metadata_ext.rs (hunk 5 above)
Comment (zjma) [reply types/src/block_metadata_ext.rs]: fixed

**Review by gregnazario — APPROVED** [review]

**Comment by larry-aptos** [comment]
1. update the description that a new field is added to API `dkg_transcript` when the validator transaction type is `dkg_result`.
2. the `changes` seem to be in the API response but in the description, it's said `omitted`. Is this expected? The events looks good.

**Comment by larry-aptos** [comment]
We can index the `quorum_certified_update ` or `dkg_transcript` fields when we have a better understanding of how these fields can be used in ecosystem projects. 

i.e. we will surface them up in protobufs but not in actual indexing.

**Review by larry-aptos — APPROVED** [review]

**Comment by zjma** [comment]
> 1. update the transaction example that a new field is added to API `dkg_transcript` when the validator transaction type is `dkg_result`. It isn't very clear to understand the example..
> 2. the `changes` seem to be in the API response but in the description, it's said `omitted`. You may want to update the example accordingly. The events look good.
> 
> Example:
> 
> image

i think i didn't want to give full example in PR description because it's too long. Perhaps I meant "detail omitted..."

**Comment by zjma** [comment]
> 1. update the transaction example that a new field is added to API `dkg_transcript` when the validator transaction type is `dkg_result`. It isn't very clear to understand the example..
> 2. the `changes` seem to be in the API response but in the description, it's said `omitted`. You may want to update the example accordingly. The events look good.
> 
> Example:
> 
> image

@larry-aptos I think i didn't want to give full example in PR description because it's too long. Perhaps I meant "detail omitted..."

File: api/types/src/convert.rs [hunk 14]
```diff
@@ -228,7 +231,9 @@ impl<'a, S: StateView> MoveConverter<'a, S> {
                     },
                 })
             },
-            ValidatorTransaction(_txn) => (info, events, timestamp).into(),
+            aptos_types::transaction::Transaction::ValidatorTransaction(txn) => {
+                Transaction::ValidatorTransaction((txn, info, events, timestamp).into())
+            },
```
Comment (zjma) [inline api/types/src/convert.rs:236]: bookmark
